File: sdks/canary/python/scenarios/d_session_title.py
#!/usr/bin/env python3
# Copyright (C) 2026 Michael Kao
# SPDX-License-Identifier: AGPL-3.0-or-later
"""D-SESSION-TITLE canary — Python SDK"""

from __future__ import annotations

import sys
import os
import time

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from canary import (
    Runner,
    Config,
    config_from_env,
    wait_active,
    ensure_session_with_retry,
)
from llmsafespaces import LLMSafeSpaces


def run(r: Runner, cfg: Config) -> None:
    if not cfg.llm_api_key:
        r.ok("session-title: skipped (no LLM API key)")
        return

    c = LLMSafeSpaces(cfg.api_url, api_key=cfg.api_key, timeout=120.0)
    ws_id = None
    try:
        ok, ws = r.assert_no_error(
            lambda: c.workspaces.create(
                name="canary-py-sess-title", runtime="base", storage_size="1Gi"
            ),
            "create: no error",
        )
        if not ok:
            return
        ws_id = ws.id

        phase = wait_active(c, ws_id)
        r.assert_(phase == "Active", "reach-active", f"got {phase!r}")
        if phase != "Active":
            return

        try:
            sess = ensure_session_with_retry(c, ws_id, 5)
        except Exception as e:
            r.fail("ensure-session: no error", str(e))
            return
        r.ok("ensure-session: no error")
        sid = sess.sessionId

        ok2, msg = r.assert_no_error(
            lambda: c.sessions.send_message(
                ws_id, sid, "What are the first 5 prime numbers? List them."
            ),
            "send-message: no error",
        )
        if ok2 and msg is not None:
            r.assert_(len(msg.content) > 0, "send-message: non-empty content")

        title_found = False
        deadline = time.time() + 20
        while time.time() < deadline:
            try:
                sessions = c.sessions.list(ws_id)
                for s in sessions:
                    if s.get("id") == sid and s.get("title"):
                        title_found = True
                        r.assert_(
                            len(s["title"]) > 0,
                            "session-title: non-empty title",
                            repr(s["title"]),
                        )
                        break
            except Exception:
                pass
            if title_found:
                break
            time.sleep(2)

        if not title_found:
            r.fail("session-title: title found within 20s", "no title found")

    finally:
        if ws_id:
            try:
                c.workspaces.delete(ws_id)
            except Exception:
                pass


if __name__ == "__main__":
    r = Runner("session-title")
    cfg = config_from_env()
    run(r, cfg)
    r.print()
    sys.exit(r.exit_code())

File: sdks/canary/typescript/scenarios/d-session-title.ts
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later
// D-SESSION-TITLE canary — TypeScript SDK

import { LLMSafeSpaces } from '../../../typescript/src/index.js';
import { Runner, Config, configFromEnv, nodeFetch, waitActive, ensureSessionWithRetry, sleep } from '../canary.js';

async function run(r: Runner, cfg: Config): Promise<void> {
  if (!cfg.llmApiKey) { r.ok('session-title: skipped (no LLM API key)'); return; }
  const c = new LLMSafeSpaces({ baseUrl: cfg.apiUrl, apiKey: cfg.apiKey, timeout: 120000, fetch: nodeFetch as any });
  let wsId: string | null = null;
  try {
    const [ok, ws] = await r.assertNoError(
      () => c.workspaces.create({ name: 'canary-ts-sess-title', runtime: 'base', storageSize: '1Gi' }),
      'create: no error');
    if (!ok || !ws) return;
    wsId = ws.id;

    const phase = await waitActive(c, wsId);
    r.assert(phase === 'Active', 'reach-active', `got "${phase}"`);
    if (phase !== 'Active') return;

    const sess = await ensureSessionWithRetry(c, wsId, 5);
    r.ok('ensure-session: no error');
    const sid = sess.sessionId;

    const [ok2, msg] = await r.assertNoError(
      () => c.sessions.sendMessage(wsId!, sid,
        'Explain in one sentence why the sky is blue during the day but orange at sunset.'),
      'send-message: no error');
    if (ok2 && msg) r.assert((msg.text ?? (msg.parts ?? []).filter(p => p.type === 'text').map(p => p.text ?? '').join('')).length > 0, 'send-message: non-empty content');

    let titleFound = false;
    const deadline = Date.now() + 20000;
    while (Date.now() < deadline && !titleFound) {
      await sleep(2000);
      const [ok3, sessions] = await r.assertNoError(
        () => c.sessions.list(wsId!), 'poll-sessions: no error');
      if (ok3 && sessions) {
        const target = sessions.find((s: any) => s.id === sid);
        if (target && target.title && target.title.length > 0) {
          titleFound = true;
          r.ok('session-title: auto-generated title present');
        }
      }
    }
    if (!titleFound) {
      r.fail('session-title: auto-generated title present', 'no title within 20s');
    }

    const [ok4, sessGet] = await r.assertNoError(
      () => c.sessions.get(wsId!, sid), 'get-session-title: no error');
    if (ok4 && sessGet) {
      r.assert(typeof (sessGet as any).title === 'string', 'get-session: has title field');
    }
  } finally {
    if (wsId) { try { await c.workspaces.delete(wsId); } catch {} }
  }
}

async function main() {
  const r = new Runner('session-title');
  await run(r, configFromEnv());
  r.print();
  process.exit(r.exitCode());
}
main().catch(e => { console.error(e); process.exit(1); });
